docs/architecture/adr-008-app-caller-cbs.md

```diff
@@ -0,0 +1,526 @@
+	// defensively guards against exploits due to incorrectly wired SendPacket ordering in IBC stacks.
+	IBCOnAcknowledgementPacketCallback(
+		cachedCtx sdk.Context,
+		packet channeltypes.Packet,
+		acknowledgement []byte,
+		relayer sdk.AccAddress,
+		contractAddress,
+		packetSenderAddress string,
+	) error
+	// IBCOnTimeoutPacketCallback is called in the source chain when a packet is not received before
+	// the timeout height. The packetSenderAddress is determined by the underlying module, and may be
+	// empty if the sender is unknown or undefined. The contract is expected to handle the callback
+	// within the user defined gas limit, and handle any error, out of gas, or panics gracefully.
+	// This entry point is called with a cached context. If an error is returned, then the changes in
+	// this context will not be persisted, but the packet lifecycle will not be blocked.
+	//
+	// Implementations are provided with the packetSenderAddress and MAY choose to use this to perform
+	// validation on the origin of a given packet. It is recommended to perform the same validation
+	// on all source chain callbacks (SendPacket, AcknowledgementPacket, TimeoutPacket). This
+	// defensively guards against exploits due to incorrectly wired SendPacket ordering in IBC stacks.
+	IBCOnTimeoutPacketCallback(
+		cachedCtx sdk.Context,
+		packet channeltypes.Packet,
+		relayer sdk.AccAddress,
+		contractAddress,
+		packetSenderAddress string,
+	) error
+	// IBCReceivePacketCallback is called in the destination chain when a packet acknowledgement is written.
+	// The contract is expected to handle the callback within the user defined gas limit, and handle any errors,
+	// out of gas, or panics gracefully.
+	// This entry point is called with a cached context. If an error is returned, then the changes in
+	// this context will not be persisted, but the packet lifecycle will not be blocked.
+	IBCReceivePacketCallback(
+		cachedCtx sdk.Context,
+		packet ibcexported.PacketI,
+		ack ibcexported.Acknowledgement,
+		contractAddress string,
+	) error
+}
+```
+
+### PacketCallbacks
+
+The packet callbacks implemented in the middleware will first call the underlying application and then route to the IBC actor callback in the post-processing step. 
+It will extract the callback data from the application packet and set the callback gas meter depending on the global limit, the user limit, and the gas left in the transaction gas meter. 
+The callback will then be routed through the callback keeper which will either panic or return a result (success or failure). In the event of a panic or an error, the callback state changes 
+are discarded and the transaction is committed.
+
+If the transaction runs out of gas because the relayer-defined limit was exceeded the entire transaction is reverted to allow for resubmission. If the chain-defined or user-defined gas limit is reached, 
+the callback state changes are reverted and the transaction is committed.
```

Comment:
The panic/revert/commit behavior is the part I need double checked specifically. @srdtrk